src/Operation/Explainable.php

```diff
@@ -0,0 +1,39 @@
+<?php
+/*
+ * Copyright 2018 MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace MongoDB\Operation;
+
+use MongoDB\Driver\Server;
+
+/**
+ * Explainable interface for explainable operations (count, distinct, group,
+ * find, findAndModify, delete, and update).
+ *
+ * @internal
+ */
+interface Explainable extends Executable
+{
+    /**
+     * Explain the operation.
+     *
+     * @param Server $server
```

Comment:
I added $server to explain() so explain can be executed inside of the operation classes.

Reply:
One issue I see with this approach is that each Explainable instance will now need to contain the same logic for executing the `explain` command on its own. That will entail quite a bit of duplicated code.

The get the most leverage out of code re-use, we can break down the many variations of `explain` into the greatest common denominator: the command document. Every variation of `explain` is going to have the same overhead for parsing the `verbosity` option and executing the explain command with `Server::executeCommand()` (`explain` doesn't take a read concern itself, so I don't think `executeReadCommand()` is needed). Then, we create an interface that can provide that common command document.

This is the design I suggested yesterday:

```php
interface Explainable {
    /* Renaming this from getCommand(), since that typically returns a
     * MongoDB\Driver\Command instance and we're more interested in
     * the document itself (Command::__construct()'s first argument). */
    function getCommandDocument();
}

class Explain implements Executable {
    /* The Explainable knows what collection it targets, so we only need
     * a database to run `explain` on. Alternatively, we might decide to also
     * pull the database from Explainable somehow, since all Operations we
     * might explain also require a database name.
     *
     * Options will at least be verbosity (the only documented explain option)
     * and typeMap, which we can apply to the cursor before the execute()
     * method returns current($cursor->toArray()) or $cursor->toArray()[0]
     * (both are equivalent). */
    public function __construct($databaseName, Explainable $explainable, array $options = []);

    public function execute(Server $server);
}
```